test/Test/Pos/Util/ModifierSpec.hs

```diff
@@ -0,0 +1,159 @@
+            prop
+                "getting the keys from a map modifier does not eliminate any from the\
+                \ produced list, unless their value was 'Nothing' or they are in the\
+                \ argument list and are a key in the map modifier simultanously"
+                (allKeysAreInMap @Text @Text)
+        describe "values" $ do
+            prop
+                "getting the values from a map modifier does not eliminate any from the\
+                \ produced list, unless their value is 'Nothing' or if it is the value\
+                \ of a key in the input list that does not belong to the the map modifier"
+                (allValsAreInMap @Text @Text)
+        describe "toList" $ do
+            prop
+                "getting the contents of the map modifier as an association list will\
+                \ keep all of the keys whose values are not 'Nothing', and will include\
+                \ the pairs in the argument list whose key is not present in the modifier"
+                (allPairsAreInAssocList @Text @Text)
+        describe "insertionsMap" $ do
+            prop
+                "the created hashmap removes all the 'Nothing' keys from the\
+                \ 'MapModifier' and keeps all the 'Just' keys"
+                (mapModifierToHashMap @Text @Text)
+        describe "insertions" $ do
+            prop
+                "the created list removes all the 'Nothing' keys from the used\
+                \ 'MapModifier' and keeps all the 'Just' keys in the association list"
+                (mapModifierToList @Text @Text)
+        describe "deletions" $ do
+            prop
+                "the created list of keys removes all those with 'Just' values and keeps\
+                \ those with 'Nothing' values"
+                (mapModifierToDeletionList @Text @Text)
+        describe "mapMaybe" $ do
+            prop
+                "the created list of keys removes all those with 'Nothing' values and\
+                \ keeps the pairs from the input list whose keys are not in the modifier"
+                (allPairsAreInMaybeList @Text @Text @Text Just)
+
+
+mapModifierFormsMonoid
+    :: (Show k, Show v, Eq k, Eq v, Hashable k)
+    => Core.MapModifier k v
+    -> Core.MapModifier k v
+    -> Core.MapModifier k v
+    -> Property
+mapModifierFormsMonoid = formsMonoid
+
+insertThenLookup
+    :: (Show k, Show v, Eq k, Eq v, Hashable k)
+    => k
+    -> v
+    -> Core.MapModifier k v
+    -> Bool
+insertThenLookup k v mapMod =
+    let newMap = Core.insert k v mapMod
+    in maybe False (== v) $ Core.lookup (const Nothing) k newMap
+
+insertDeleteThenLookup
+    :: (Show k, Show v, Eq k, Eq v, Hashable k)
+    => k
+    -> v
+    -> Core.MapModifier k v
+    -> Bool
+insertDeleteThenLookup k v mapMod =
+    let newMap@(Core.MapModifier mm) = Core.insert k v mapMod
+        (Core.MapModifier dm) = Core.delete k newMap
+    in (HM.lookup k mm == Just (Just v)) && (HM.lookup k dm == Just Nothing)
```

Comment:
> The issue is, I didn't want the tests to depend on the functions I'm testing. What if there's a mistake in insert that causes delete's test to wrongly pass?

You shouldn't depend on the functions you're testing. You know some invariants about these functions and expect them to hold. For instance, you expect that if you `insert` something into `MapModifier` then subsequent lookup will return this value regardless of contents of underlying map. You also know that if you `delete` something from `MapModifier` then subsequent lookup will return `Nothing` regardless of contents on underlying map.

These are results which user of `MapModifier` expects. However, user of `MapModifier` doesn't care about internal representation of `MapModifier`. The fact that `(HM.lookup k mm == Just (Just v))` doesn't mean anything for the person not familiar with implementation of `MapModifier`.

`MapModifier` is quite small and simple structure, so this example is not really representable, but I hope you got the idea.

About this question:
> What if there's a mistake in insert that causes delete's test to wrongly pass?

Well, yes, it can happen that `insert` does nothing and `delete` does nothing and as a result combination of `insert` and `delete` works as expected. But in this case neither `insert` nor `delete` work as expected in isolation. That's why there should be tests for these functions in isolation and for combination.